src/Microsoft.DotNet.XHarness.CLI/Commands/WASM/WasmBrowserTestRunner.cs

```diff
@@ -0,0 +1,190 @@
+            }
+        }
+
+        private async Task RunLogMessagesPump(IWebDriver driver, TaskCompletionSource<ExitCode> taskCompletion, CancellationToken token)
+        {
+            try
+            {
+                while (!token.IsCancellationRequested && !taskCompletion.Task.IsCompleted)
+                {
+                    foreach (var logEntry in driver.Manage().Logs.GetLog(LogType.Browser))
+                    {
+                        if (logEntry.Level == SLogLevel.Severe)
+                        {
+                            // throw on driver errors, or other errors that show up
+                            // in the console log
+                            throw new ArgumentException(logEntry.Message);
+                        }
+
+                        var match = s_consoleLogRegex.Value.Match(Regex.Unescape(logEntry.Message));
+                        string msg = match.Success ? match.Groups[1].Value : logEntry.Message;
+                        msg += Environment.NewLine;
+
+                        _processLogMessage(msg);
+
+                        if (!msg.StartsWith("WASM EXIT "))
+                        {
+                            continue;
+                        }
+
+                        if (int.TryParse(msg.Substring("WASM EXIT ".Length).Trim(), out var code))
+                        {
+                            taskCompletion.SetResult((ExitCode) Enum.ToObject(typeof(ExitCode), code));
+                        }
+                        else
+                        {
+                            _logger.LogDebug($"Got an unknown exit code in msg: '{msg}'");
+                            taskCompletion.SetResult(ExitCode.TESTS_FAILED);
+                        }
+                    }
+
+                    if (driver.CurrentWindowHandle == null) {
+                        // Accessing this property will throw, if chrome has crashed
+                    }
+
+                    await Task.Delay(100, token).ConfigureAwait(false);
+                }
+            }
+            catch (TaskCanceledException)
+            {
+                // Ignore
+            }
+        }
+
+        private string BuildUrl(string webServerAddr)
+        {
+            var uriBuilder = new UriBuilder($"{webServerAddr}/index.html");
+            var sb = new StringBuilder();
+            foreach (var arg in _passThroughArguments)
+            {
+                if (sb.Length > 0)
+                    sb.Append("&");
+
+                sb.Append($"arg={HttpUtility.UrlEncode(arg)}");
+            }
+
+            uriBuilder.Query = sb.ToString();
+            return uriBuilder.ToString();
+        }
+
+        private static async Task<string> StartWebServer(CancellationToken token)
+        {
+            var host = new WebHostBuilder()
+                .UseSetting("UseIISIntegration", false.ToString())
+                .UseKestrel()
+                .UseContentRoot(Directory.GetCurrentDirectory())
```

Comment:
Instead of using `Directory.GetCurrentDirectory()` we should flow the `_arguments.OutputDirectory` somehow here and create a folder there. This makes it more predictable in case the working directory of the process is not what we expect.

Reply:
Isn't this where the web server gets files to serve from? Then I guess you would want the current folder with the files that you included in the payload of the work item?

I would also be careful with `_arguments.OutputDirectory` - this directory gets picked up by Helix and files inside uploaded with the results.
`Directory.GetCurrentDirectory()` is probably the Helix working directory which gets deleted after job is over so makes sense for that flow but not sure what this will cause for local development.

Reply:
Here, the outputDirectory seems to be set to `artifacts/bin/System.Collections.Immutable.Tests/net5.0-release/browser-wasm/AppBundle/xharness-output`, so that just gets the test results xml file. But the files that we need to host are one directory up.

Reply:
yeah my mental model of where the served files come from was broken :)
I still think we should pass the path to the files into xharness rather than relying on `Directory.GetCurrentDirectory()` being the right thing.

Basically pass the directory in like we do for `_arguments.JSFile`.

Reply:
Can I use `--app, -a=VALUE        Path to already-packaged app`, (`AppPackagePath`)? It doesn't seem to have a default value though, and is not a required parameter. I could use curdir as the default for this.

Reply:
To be clear, that's an existing parameter

Reply:
Err though the property throws if it isn't set (https://github.com/dotnet/xharness/blob/master/src/Microsoft.DotNet.XHarness.CLI/CommandArguments/TestCommandArguments.cs#L26). And making it required will probably break existing stuff.

Reply:
Hm yeah we should probably move `--app` and `--targets` down into Android/iOSTestCommandArguments.cs since they don't really make sense on wasm, but that can be done in a future PR :)